File: packages/gbnf-experiment/data/20260909T144331Z_5b46f630/ported_implementation/src/grammarGraph/grammarGraphTypes.ts
import type { Pointers } from './pointers.js';
import type { RuleRef } from './ruleRef.js';

export type Colorize = (text: string | number, color: string) => string;

export interface PrintOpts {
  pointers?: Pointers;
  colorize: Colorize;
  show_position?: boolean;
}

export type Range = [number, number];

export type RuleDict = { type: string; value?: unknown };

export abstract class Rule {
  abstract get type(): string;

  equals(other: unknown): boolean {
    return other instanceof Rule && other.constructor === this.constructor;
  }

  toDict(): RuleDict {
    return { type: this.type };
  }

  toJSON(): RuleDict {
    return this.toDict();
  }

  toString(): string {
    return `${this.type}()`;
  }
}

export abstract class RuleWithValue<T> extends Rule {
  value: T;

  constructor(value: T) {
    super();
    this.value = value;
  }

  override equals(other: unknown): boolean {
    return (
      other instanceof RuleWithValue &&
      other.constructor === this.constructor &&
      JSON.stringify(this.value) === JSON.stringify(other.value)
    );
  }

  override toDict(): RuleDict {
    return { ...super.toDict(), value: this.value };
  }

  override toString(): string {
    return `${this.type}(value=${JSON.stringify(this.value)})`;
  }
}

export abstract class RuleWithListOfIntsOrRanges extends RuleWithValue<(number | Range)[]> {
  constructor(value: (number | Range)[] = []) {
    // the reference implementation copies the incoming list so that later mutation of the
    // rule's value cannot be observed by the caller.
    super([...value]);
  }
}

export class RuleChar extends RuleWithListOfIntsOrRanges {
  get type(): 'RuleChar' {
    return 'RuleChar';
  }
}

export class RuleCharExclude extends RuleWithListOfIntsOrRanges {
  get type(): 'RuleCharExclude' {
    return 'RuleCharExclude';
  }
}

export class RuleEnd extends Rule {
  get type(): 'RuleEnd' {
    return 'RuleEnd';
  }
}

export type UnresolvedRule = RuleChar | RuleCharExclude | RuleRef | RuleEnd;

// ValidInput can either be a string, or a number indicating a code point.
// It CANNOT be a number representing a number; a number intended as input (like "8")
// should be passed in as a string.
export type ValidInput = string | number | number[];

// RuleRefs should never be exposed to the end user.
export type ResolvedRule = RuleCharExclude | RuleChar | RuleEnd;

File: proposed-projects/round-trip-experiment/reverse/20260909T144331Z_5b46f630/20260910T083554Z_f6f6274b/ported_implementation/gbnf/grammar_graph/grammar_graph_types.py
import json
from typing import Any, Callable, Dict, List, Optional, Sequence, Union

Colorize = Callable[[Union[str, int], str], str]

# A range is a two element list of code points, [low, high].
Range = List[int]

RuleDict = Dict[str, Any]

# ValidInput can either be a string, or a number indicating a code point.
# It CANNOT be a number representing a number; a number intended as input (like "8")
# should be passed in as a string.
ValidInput = Union[str, int, List[int]]


class PrintOpts:
    def __init__(
        self,
        colorize: Colorize,
        pointers: Optional[Any] = None,
        show_position: bool = False,
    ):
        self.colorize = colorize
        self.pointers = pointers
        self.show_position = show_position


class Rule:
    @property
    def type(self) -> str:
        raise NotImplementedError

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Rule) and type(other) is type(self)

    def __hash__(self) -> int:
        return hash(self.type)

    def to_dict(self) -> RuleDict:
        return {"type": self.type}

    def to_json(self) -> RuleDict:
        return self.to_dict()

    def __repr__(self) -> str:
        return f"{self.type}()"


class RuleWithValue(Rule):
    def __init__(self, value: Any):
        self.value = value

    def __eq__(self, other: object) -> bool:
        return (
            isinstance(other, RuleWithValue)
            and type(other) is type(self)
            and self.value == other.value
        )

    def __hash__(self) -> int:
        return hash((self.type, json.dumps(self.value)))

    def to_dict(self) -> RuleDict:
        return {**super().to_dict(), "value": self.value}

    def __repr__(self) -> str:
        return f"{self.type}(value={json.dumps(self.value)})"


class RuleWithListOfIntsOrRanges(RuleWithValue):
    def __init__(self, value: Optional[Sequence[Union[int, Range]]] = None):
        # copy the incoming list so that later mutation of the rule's value cannot be
        # observed by the caller.
        super().__init__(list(value) if value is not None else [])


class RuleChar(RuleWithListOfIntsOrRanges):
    @property
    def type(self) -> str:
        return "RuleChar"


class RuleCharExclude(RuleWithListOfIntsOrRanges):
    @property
    def type(self) -> str:
        return "RuleCharExclude"


class RuleEnd(Rule):
    @property
    def type(self) -> str:
        return "RuleEnd"
